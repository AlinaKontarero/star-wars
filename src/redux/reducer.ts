import { Reducer } from 'redux';
import {
  CharacterActions,
  PersonActionTypes,
} from './actions';
import { IPerson } from './types';


export interface IPersonstate {
  readonly Persons: IPerson[];
}

export const initialPersonstate: IPersonstate = {
  Persons: [],
};

export const personReducer: Reducer<IPersonstate, CharacterActions> = (
  state = initialPersonstate,
  action
) => {
  switch (action.type) {
    case PersonActionTypes.GET_ALL: {
      return {
        ...state,
        Persons: action.Persons,
      };
    }
    default:
      return state;
  }
};