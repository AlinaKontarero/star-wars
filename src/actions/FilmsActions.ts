import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import axios from 'axios';
import { IFilm, IFilmState } from '../reducers/filmReducer';

export enum FilmActionTypes {
  GET_ALL = 'GET_ALL',
}

export interface IFilmGetAllAction {
  type: FilmActionTypes.GET_ALL;
  Films: IFilm[];
}

export type FilmActions = IFilmGetAllAction;

/* Get All Action
<Promise<Return Type>, State Interface, Type of Param, Type of Action> */
export const getAllFilms: ActionCreator<
  ThunkAction<Promise<void>, IFilmState, null, IFilmGetAllAction>
> = () => {
  return async (dispatch: Dispatch) => {
    try {
      const response = await axios.get('https://swapi.dev/api/films');
      dispatch({
        Films: response.data.results,
        type: FilmActionTypes.GET_ALL,
      });
    } catch (err) {
      console.error(err);
    }
  };
};