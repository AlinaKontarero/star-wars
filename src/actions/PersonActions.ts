// Import redux types
import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import axios from 'axios';

// Import Character Typing
import { ICharacter, IPersonstate } from '../reducers/personReducer';

// Create Action Constants
export enum CharacterActionTypes {
  GET_ALL = 'GET_ALL',
}

// Interface for Get All Action Type
export interface ICharacterGetAllAction {
  type: CharacterActionTypes.GET_ALL;
  Persons: ICharacter[];
}

/* 
Combine the action types with a union (we assume there are more)
example: export type CharacterActions = IGetAllAction | IGetOneAction ... 
*/
export type CharacterActions = ICharacterGetAllAction;

/* Get All Action
<Promise<Return Type>, State Interface, Type of Param, Type of Action> */
export const getAllPersons: ActionCreator<
  ThunkAction<Promise<any>, IPersonstate, null, ICharacterGetAllAction>
> = () => {
  return async (dispatch: Dispatch) => {
    try {
      const response = await axios.get('https://swapi.dev/api/people');
      dispatch({
        Persons: response.data.results,
        type: CharacterActionTypes.GET_ALL,
      });
    } catch (err) {
      console.error(err);
    }
  };
};