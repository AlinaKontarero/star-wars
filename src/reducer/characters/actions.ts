import { Action, Dispatch } from "redux";

export const SET_CHARACTERS = 'SET_CHARACTERS';

export const getCharacters = () => {
  return (dispatch: Dispatch<Action>) =>
    fetch('https://swapi.dev/api/people')
      .then(res => res.json())
      .then(res => res.results)
      .then(characters =>
        dispatch(setCharacters(characters))
      );
}

export const setCharacters = (characters: any[]) => {
  return {
    type: SET_CHARACTERS,
    characters,
  };
}