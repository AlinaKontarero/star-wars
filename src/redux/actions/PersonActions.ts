import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import axios from 'axios';

import { IPerson, IPersonstate } from '../reducers/personReducer';

export enum PersonActionTypes {
  GET_ALL = 'GET_ALL',
}
export interface IPersonGetAllAction {
  type: PersonActionTypes.GET_ALL;
  Persons: IPerson[];
}

export type CharacterActions = IPersonGetAllAction;

export const getAllPersons: ActionCreator<
  ThunkAction<Promise<any>, IPersonstate, null, IPersonGetAllAction>
> = () => {
  return async (dispatch: Dispatch) => {
    try {
      const reqLinks = [];
      const testCall = await axios.get('https://swapi.dev/api/people')
      const pagesNumber: number = Math.ceil(testCall.data.count / testCall.data.results.length)
      // Start from second page as we already have the data from 1st page:
      for (let i = 2; i <= pagesNumber; i++) {
        reqLinks.push('https://swapi.dev/api/people/?page=' + i)
      }

      const responses = await Promise.all(
        reqLinks.map(link => 
          axios(link).then(({data: responses}) => responses)
        ))
      
      const result = [...responses, testCall.data]
        .map(_res => _res.results) 
        .flat(1)

      dispatch({
        Persons: result,
        type: PersonActionTypes.GET_ALL,
      });
    } catch (err) {
      console.error(err);
    }
  };
};