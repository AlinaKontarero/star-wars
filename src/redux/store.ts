import { applyMiddleware, combineReducers, createStore, Store } from 'redux';
import thunk from 'redux-thunk';
import {
  personReducer,
  IPersonstate,
} from './reducer';
export interface IAppState {
  PersonState: IPersonstate;
}

const rootReducer = combineReducers<IAppState>({
  PersonState: personReducer,
});

export default function configureStore(): Store<IAppState, any> {
  const store = createStore(rootReducer, undefined, applyMiddleware(thunk));
  return store;
}