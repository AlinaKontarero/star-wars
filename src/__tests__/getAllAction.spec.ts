import configureMockStore from 'redux-mock-store'
import { PersonActionTypes } from '../redux/actions';
import { IPerson } from '../redux/types';
import { mockPersons } from './mockups/mockPersons';

const getPersons = (persons: IPerson[]) => ({
  type: PersonActionTypes.GET_ALL,
  Persons: persons
});

const mockStore = configureMockStore()
const store = mockStore({})
describe('action GET_ALL', () => {
  it('creates Persons array', () => {
    store.dispatch(getPersons(mockPersons));
    expect(store.getActions()).toMatchSnapshot();
  });
});