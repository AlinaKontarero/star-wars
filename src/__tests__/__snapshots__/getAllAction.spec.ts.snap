// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action GET_ALL creates Persons array 1`] = `
Array [
  Object {
    "Persons": Array [
      Object {
        "birth_year": "19BBY",
        "created": "2014-12-09T13:50:51.644000Z",
        "edited": "2014-12-20T21:17:56.891000Z",
        "eye_color": "blue",
        "films": Array [
          "http://swapi.dev/api/films/1/",
          "http://swapi.dev/api/films/2/",
          "http://swapi.dev/api/films/3/",
          "http://swapi.dev/api/films/6/",
        ],
        "gender": "male",
        "hair_color": "blond",
        "height": "172",
        "homeworld": "http://swapi.dev/api/planets/1/",
        "mass": "77",
        "name": "Luke Skywalker",
        "skin_color": "fair",
        "species": Array [],
        "starships": Array [
          "http://swapi.dev/api/starships/12/",
          "http://swapi.dev/api/starships/22/",
        ],
        "url": "http://swapi.dev/api/people/1/",
        "vehicles": Array [
          "http://swapi.dev/api/vehicles/14/",
          "http://swapi.dev/api/vehicles/30/",
        ],
      },
      Object {
        "birth_year": "112BBY",
        "created": "2014-12-10T15:10:51.357000Z",
        "edited": "2014-12-20T21:17:50.309000Z",
        "eye_color": "yellow",
        "films": Array [
          "http://swapi.dev/api/films/1/",
          "http://swapi.dev/api/films/2/",
          "http://swapi.dev/api/films/3/",
          "http://swapi.dev/api/films/4/",
          "http://swapi.dev/api/films/5/",
          "http://swapi.dev/api/films/6/",
        ],
        "gender": "n/a",
        "hair_color": "n/a",
        "height": "167",
        "homeworld": "http://swapi.dev/api/planets/1/",
        "mass": "75",
        "name": "C-3PO",
        "skin_color": "gold",
        "species": Array [
          "http://swapi.dev/api/species/2/",
        ],
        "starships": Array [],
        "url": "http://swapi.dev/api/people/2/",
        "vehicles": Array [],
      },
    ],
    "type": "GET_ALL",
  },
]
`;
